File: backend/src/main/java/lan/dk/podcastserver/manager/downloader/RTMPDownloader.java
package lan.dk.podcastserver.manager.downloader;


import com.github.davinkevin.podcastserver.service.MimeTypeService;
import com.github.davinkevin.podcastserver.service.ProcessService;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Status;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.repository.PodcastRepository;
import lan.dk.podcastserver.service.properties.ExternalTools;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Scope;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;

import java.io.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static io.vavr.API.Try;
import static java.util.Objects.nonNull;

@Slf4j
@Component("RTMPDownloader")
@Scope("prototype")
public class RTMPDownloader extends AbstractDownloader {

    private final ProcessService processService;
    private final ExternalTools externalTools;

    int pid = 0;
    Process p = null;

    public RTMPDownloader(ItemRepository itemRepository, PodcastRepository podcastRepository, PodcastServerParameters podcastServerParameters, SimpMessagingTemplate template, MimeTypeService mimeTypeService, ProcessService processService, ExternalTools externalTools) {
        super(itemRepository, podcastRepository, podcastServerParameters, template, mimeTypeService);
        this.processService = processService;
        this.externalTools = externalTools;
    }

    @Override
    public Item download() {
        log.debug("Download");

        try {
            target = getTargetFile(item);
            log.debug("Fichier de sortie : {}" , target.toAbsolutePath().toString());

            p  = processService
                    .newProcessBuilder(externalTools.getRtmpdump(), "-r", getItemUrl(item), "-o", target.toAbsolutePath().toString())
                    .directory(new File("/tmp"))
                    .redirectErrorStream(true)
                    .start();

            pid = processService.pidOf(p);

            new RTMPWatcher(this).run();

            p.waitFor();
            pid = 0;
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
        return item;
    }

    @Override
    public void startDownload() {
        if (pid != 0 && nonNull(p)) { //Relancement du process UNIX
            log.debug("Stop previous process");
            p.destroy();
        }
        super.startDownload();
    }

    @Override
    public void pauseDownload() {
        ProcessBuilder stopProcess = processService.newProcessBuilder("kill", "-STOP", "" + pid);
        Try(stopProcess::start)
            .andThen(super::pauseDownload)
            .onFailure(e -> {
                log.error("IOException :", e);
                this.stopDownload();
            });
    }

    @Override
    public void stopDownload() {
        destroyProcess();
        super.stopDownload();
    }

    private void destroyProcess() {
        if (nonNull(p)) {
            p.destroy();
        }
    }

    @Override
    public void failDownload() {
        destroyProcess();
        super.failDownload();
    }

    @Override
    public Integer compatibility(DownloadingItem ditem) {
        return ditem.getUrls().length() == 1 && ditem.getUrls().head().startsWith("rtmp://")
                ? 1
                : Integer.MAX_VALUE;
    }

    @Slf4j
    @RequiredArgsConstructor
    static class RTMPWatcher implements Runnable {

        static final String DOWNLOAD_COMPLETE = "download complete";
        static final Pattern RTMPDUMP_PROGRESSION_PATTERN_EXTRACTOR = Pattern.compile("[^(]*\\(([0-9]*).*%\\)");
        final RTMPDownloader rtmpDownloader;

        private BufferedReader getBufferedReader(InputStream is) {
            return new BufferedReader(new InputStreamReader(is));
        }

        @Override
        public void run() {
            Process p = rtmpDownloader.p;
            Item item = rtmpDownloader.item;
            Integer pid = rtmpDownloader.pid;

            BufferedReader br = getBufferedReader(p.getInputStream());
            String line;
            log.debug("Reading output of RTMPDump");
            try {
                while ((line = br.readLine()) != null) {
                    log.debug(line);
                    Matcher m = RTMPDUMP_PROGRESSION_PATTERN_EXTRACTOR.matcher(line);
                    if (progressionHasChange(item, m)) {
                        item.setProgression(Integer.parseInt(m.group(1)));
                        rtmpDownloader.convertAndSaveBroadcast();
                    } else if (isDownloadComplete(line)) {
                        log.info("End of download");
                        rtmpDownloader.finishDownload();
                        break;
                    }
                    if (pid == 0) {
                        break;
                    }
                }
            } catch (IOException e) {
                log.error("IOException :", e);
            }
            if (Status.FINISH != item.getStatus() && !rtmpDownloader.stopDownloading.get()) {
                throw new RuntimeException("Unexpected ending, failed download");
            }
        }

        private boolean isDownloadComplete(String ligne) {
            return ligne.toLowerCase().contains(DOWNLOAD_COMPLETE);
        }

        private boolean progressionHasChange(Item item, Matcher m) {
            return m.matches() && Integer.parseInt(m.group(1)) > item.getProgression();
        }
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/downloader/RTMPDownloader.kt
package com.github.davinkevin.podcastserver.manager.downloader


import com.github.davinkevin.podcastserver.download.DownloadRepository
import com.github.davinkevin.podcastserver.entity.Status
import com.github.davinkevin.podcastserver.messaging.MessagingTemplate
import com.github.davinkevin.podcastserver.service.ProcessService
import com.github.davinkevin.podcastserver.service.properties.ExternalTools
import com.github.davinkevin.podcastserver.service.storage.FileStorageService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import java.io.File
import java.time.Clock
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.regex.Pattern

@Component
@Scope(SCOPE_PROTOTYPE)
class RTMPDownloader(
    downloadRepository: DownloadRepository,
    template: MessagingTemplate,
    clock: Clock,
    file: FileStorageService,
    val processService: ProcessService,
    val externalTools: ExternalTools
) : AbstractDownloader(downloadRepository, template, clock, file) {

    private val log = LoggerFactory.getLogger(RTMPDownloader::class.java)

    internal var pid = 0L
    private var p: Process? = null
    private val stopDownloading = AtomicBoolean(false)

    override fun download(): DownloadingItem {
        log.debug("Download")

        try {
            target = computeTargetFile(downloadingInformation)
            log.debug("out file : {}", target.toAbsolutePath().toString())

            val processToExecute = processService
                .newProcessBuilder(
                    externalTools.rtmpdump,
                    "-r",
                    downloadingInformation.url(),
                    "-o",
                    target.toAbsolutePath().toString()
                )
                .directory(File("/tmp"))
                .redirectErrorStream(true)

            p = processService.start(processToExecute)
            pid = processService.pidOf(p!!)

            listenLogs(p!!)

            p!!.waitFor()
            pid = 0
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        return downloadingInformation.item
    }

    fun listenLogs(process: Process) {
        log.debug("Reading output of RTMPDump")

        process.inputStream
            .bufferedReader()
            .lines()
            .forEach {
                val (isProgression, progression) = isProgressionLine(it)
                if (isProgression) {
                    broadcastProgression(downloadingInformation.item, progression)
                    return@forEach
                }

                if (isDownloadComplete(it)) {
                    log.info("End of download")
                    finishDownload()
                    return@forEach
                }
            }

        if (downloadingInformation.item.status != Status.FINISH && !stopDownloading.get()) {
            throw RuntimeException("Unexpected ending, failed download")
        }
    }

    override fun startDownload() {
        stopDownloading.set(false)
        super.startDownload()
    }

    override fun stopDownload() {
        stopDownloading.set(true)
        destroyProcess()
        super.stopDownload()
    }

    private fun destroyProcess() {
        if (p != null) p!!.destroy()
    }

    override fun failDownload() {
        stopDownloading.set(true)
        destroyProcess()
        super.failDownload()
    }

    override fun compatibility(downloadingInformation: DownloadingInformation) =
        if (downloadingInformation.urls.size == 1 && downloadingInformation.urls.first().startsWith("rtmp://")) 1
        else Integer.MAX_VALUE

    private fun broadcastProgression(item: DownloadingItem, progression: Int) {
        if (item.progression != progression)
            broadcast(downloadingInformation.item)
    }

    private fun isProgressionLine(line: String): Pair<Boolean, Int> {
        val m = RTMP_DUMP_PROGRESSION.matcher(line)
        val isMatch = m.matches()
        return if (isMatch) true to m.group(1).toInt()
        else false to -1
    }

    private fun isDownloadComplete(line: String) = line.lowercase(Locale.getDefault()).contains(DOWNLOAD_COMPLETE)

    companion object {
        private const val DOWNLOAD_COMPLETE = "download complete"
        private val RTMP_DUMP_PROGRESSION = Pattern.compile("[^(]*\\(([0-9]*).*%\\)")!!
    }
}
